drafts/arcalinea-tx-expiry/draft.rst

```diff
@@ -0,0 +1,61 @@
+**Title:** Transaction Expiry
+
+**Author:** Jay Graber
+
+**Status:** Active
+
+**Category:** Standards
+
+**Created:** 2018-01-09
+
+Abstract
+===========
+
+This is an Standards ZIP describing a new consensus rule to expire transactions that are not mined from the mempool.
+
+Motivation
+===========
+
+Transactions that are too large or have insufficient fees are often not mined. This indeterminism is a source of confusion for users and wallets. Giving users the ability to specify the number of blocks a transaction should sit in the mempool before it expires, and otherwise setting a default, would provide certainty around how long a transaction has to confirm.
+
+Advantages include improving performance by removing transactions unlikely to be mined, and simplifying bidirectional payment channels by reducing the need to store and compress revocations for past states since transactions are allowed to expire.
+
+Specification
+===============
+
+Transactions will have a new field, nBlockExpiry, which will set the blockheight after which transactions will be removed from the mempool if they have not been mined. The last block that the transaction below could possibly be included in is 3539. After that, it will be removed from the mempool.
+
+````
+"txid": "17561b98cc77cd5a984bb959203e073b5f33cf14cbce90eb32b95ae2c796723f",
+"version": 3,
+"locktime": 2089,
+"blockexpiry": 3539,
+```
+
+With the addition of this feature, zero-confirmation transactions with an expiration blockheight set will have even less guarantee of inclusion. This means that UIs and services must never rely on zero-confirmation transactions in Zcash.
+
+Default behavior
+-----------------
+
+Default behavior does not change -- transactions do not expire. (Alternative proposal is to set a default expiry period of 2 days, ~1440 blocks.) When nBlockExpiry is set to 0, the transaction will not expire from the mempool.
+
+If used in combination with nLockTime, nLockTime must be a blockheight, and nBlockExpiry must be higher. To simplify this feature, we could disallow its use in conjunction with nLockTime.
```

Comment:
IIRC modern Bitcoin wallets set nLockTime to the most recent block to protect against some particular attack, so if we pulled that in (which we might not, or might do differently in order to sync it with the anchor selection), then users wouldn't be able to use `nBlockExpiry`.

We can just enforce in consensus that valid transactions have `nLockTime < nBlockExpiry` (and also correct relative units, if we add support for time-expiry too.

Reply:
The default behavior is to set nLockTime to current block - 10. I think it's fine to use in conjunction with nLockTime as long as both are blockheights; @bitcartel suggested disallowing its use with nLocktime.